File: src-tauri/src/models/patient.rs
use crate::models::alert::PatientAlert;
use crate::models::dispensing::DispensingRecord;
use crate::models::treatment::{Followup, Outcome, TreatmentPlan};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone, sqlx::FromRow)]
pub struct TbPatient {
  pub id: i64,
  pub hn: String,
  pub enrolled_at: String,
  pub enrolled_by: Option<String>,
  pub status: String,
  pub tb_type: Option<String>,
  pub diagnosis_date: Option<String>,
  pub notes: Option<String>,
  pub created_at: String,
  pub updated_at: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct PatientDemographics {
  pub hn: String,
  pub full_name: String,
  pub age: Option<i64>,
  pub sex: Option<String>,
  pub address: Option<String>,
  pub phone: Option<String>,
  pub birthday: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct PatientDrugRecord {
  pub hn: String,
  pub full_name: String,
  pub age: Option<i64>,
  pub sex: Option<String>,
  pub first_dispensed: Option<String>,
  pub last_dispensed: Option<String>,
  pub visit_count: i64,
  pub drug_names: Option<String>,
  pub drug_classes: Vec<String>,
  pub is_enrolled: bool,
  pub patient_status: Option<String>,
}

use crate::models::settings::RegimenPhase;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct EnrollmentInput {
  pub hn: String,
  pub tb_type: String,
  pub diagnosis_date: Option<String>,
  pub regimen: String,
  pub treatment_start_date: String,
  pub enrolled_by: Option<String>,
  pub notes: Option<String>,
  /// Optional pre-resolved phase definitions. When `None`, the backend
  /// will attempt to look up the regimen name in `regimen_definitions`
  /// settings, or fall back to parsing the regimen string.
  #[serde(default)]
  pub regimen_phases: Option<Vec<RegimenPhase>>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct PatientDetail {
  pub patient: TbPatient,
  pub demographics: Option<PatientDemographics>,
  pub current_plan: Option<TreatmentPlan>,
  pub followups: Vec<Followup>,
  pub outcome: Option<Outcome>,
  pub dispensing_history: Vec<DispensingRecord>,
  pub alerts: Vec<PatientAlert>,
  /// Whether the HOSxP MySQL connection was available when this detail was loaded.
  pub mysql_connected: bool,
  /// Error message from the MySQL side (demographics or dispensing fetch), if any.
  pub mysql_error: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ActivePatientRow {
  pub tb_patient: TbPatient,
  pub demographics: Option<PatientDemographics>,
  pub current_plan: Option<TreatmentPlan>,
  pub current_month: Option<i64>,
  pub total_months: Option<i64>,
  pub days_since_last_dispensing: Option<i64>,
  pub outcome_value: Option<String>,
  pub alerts: Vec<PatientAlert>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct SearchFilters {
  pub date_from: Option<String>,
  pub date_to: Option<String>,
  pub drug_classes: Option<Vec<String>>,
  pub enrollment_status: Option<String>,
  pub hn_search: Option<String>,
  pub name_search: Option<String>,
  pub page: Option<i64>,
  pub page_size: Option<i64>,
}

/// A single TB clinic appointment record fetched from HOSxP `oapp` (clinic = '009').
#[derive(Debug, Serialize, Deserialize, Clone, sqlx::FromRow)]
pub struct AppointmentRecord {
  pub hn: String,
  pub full_name: Option<String>,
  /// Appointment date as `YYYY-MM-DD`.
  pub nextdate: String,
}

File: src/types/patient.ts
export interface TbPatient {
  id: number;
  hn: string;
  enrolled_at: string;
  enrolled_by: string | null;
  status: 'active' | 'completed' | 'transferred' | 'died' | 'defaulted';
  tb_type: 'pulmonary' | 'extra_pulmonary' | null;
  diagnosis_date: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface PatientDemographics {
  hn: string;
  full_name: string;
  age: number | null;
  sex: string | null;
  address: string | null;
  phone: string | null;
  birthday: string | null;
}

export interface PatientDrugRecord {
  hn: string;
  full_name: string;
  age: number | null;
  sex: string | null;
  first_dispensed: string | null;
  last_dispensed: string | null;
  visit_count: number;
  drug_names: string | null;
  drug_classes: string[];
  is_enrolled: boolean;
  patient_status?: string | null;
}

export interface EnrollmentInput {
  hn: string;
  tb_type: string;
  diagnosis_date: string | null;
  regimen: string;
  treatment_start_date: string;
  enrolled_by: string | null;
  notes: string | null;
}

export interface PatientDetail {
  patient: TbPatient;
  demographics: PatientDemographics | null;
  current_plan: import('./treatment').TreatmentPlan | null;
  followups: import('./treatment').Followup[];
  outcome: import('./treatment').Outcome | null;
  dispensing_history: import('./dispensing').DispensingRecord[];
  alerts: import('./alert').PatientAlert[];
  mysql_connected: boolean;
  mysql_error: string | null;
}

export interface ActivePatientRow {
  tb_patient: TbPatient;
  demographics: PatientDemographics | null;
  current_plan: import('./treatment').TreatmentPlan | null;
  current_month: number | null;
  total_months: number | null;
  days_since_last_dispensing: number | null;
  outcome_value: string | null;
  alerts: import('./alert').PatientAlert[];
}

export interface SearchFilters {
  date_from?: string;
  date_to?: string;
  drug_classes?: string[];
  enrollment_status?: 'all' | 'enrolled' | 'not_enrolled' | 'discharged';
  page?: number;
  page_size?: number;
  hn_search?: string;
  name_search?: string;
}
